tests/client/disco/containers/TestInstallButton.js

```diff
@@ -21,152 +11,207 @@ import {
+      button.statusChanged({state: 'STATE_INSTALLED'});
+      assert(dispatch.calledWith({
+        type: 'INSTALL_COMPLETE',
+        payload: {slug},
+      }));
+    });
+
+    describe('mapStateToProps', () => {
+      it('pulls the installation data from the state', () => {
+        const addon = {
+          slug: 'addon',
+          downloadProgress: 75,
+        };
+        assert.deepEqual(
+          mapStateToProps({
+            installations: {foo: {some: 'data'}, addon},
+            addons: {addon: {addonProp: 'addonValue'}},
+          }, {slug: 'addon'}),
+          {slug: 'addon', downloadProgress: 75, addonProp: 'addonValue'});
+      });
+
+      it('handles no installation data', () => {
+        assert.deepEqual(mapStateToProps({}, {slug: 'foo'}), {});
+      });
+    });
+  });
+
+  describe('using manager', () => {
+    it('should call install function on click when uninstalled', () => {
+      const dispatch = sinon.spy();
+      const manager = stubAddonManager();
+      const slug = 'foo';
+      manager.install = sinon.spy();
+      const button = renderButton({dispatch, slug, status: UNINSTALLED});
       const root = findDOMNode(button);
-      const checkbox = root.querySelector('input[type=checkbox]');
-      assert.equal(checkbox.hasAttribute('disabled'), true);
-      assert.ok(root.classList.contains('unknown'));
+      Simulate.click(root);
+      assert(manager.install.called);
+      assert(dispatch.calledWith({
+        type: 'START_DOWNLOAD',
+        payload: {slug},
+      }));
     });
 
-    it('pulls the installation data from the state', () => {
-      const addon = {
-        slug: 'addon',
-        downloadProgress: 75,
-      };
-      assert.strictEqual(
-        mapStateToProps(
-          {installations: {foo: {some: 'data'}, addon}},
-          {slug: 'addon'}),
-        addon);
+    it('should update the status to INSTALLED in componentDidMount', () => {
+      const dispatch = sinon.spy();
+      const manager = stubAddonManager();
+      const guid = '@foo';
+      const slug = 'foo';
+      const installURL = 'http://the.url';
+      renderButton({dispatch, guid, slug, status: UNKNOWN, installURL});
+      return manager.getAddon().then(() => {
+        assert(dispatch.calledWith({
+          type: 'INSTALL_STATE',
+          payload: {guid, slug, status: INSTALLED, url: installURL},
+        }));
+      });
     });
 
-    it('handles no installation data', () => {
-      assert.deepEqual(mapStateToProps({}, {slug: 'foo'}), {});
+    it('should update the status to UNINSTALLED in componentDidMount', () => {
+      const dispatch = sinon.spy();
+      const manager = stubAddonManager({getAddon: Promise.reject()});
+      const guid = '@foo';
+      const slug = 'foo';
+      const installURL = 'http://the.url';
+      renderButton({dispatch, guid, slug, status: UNKNOWN, installURL});
+      return manager.getAddon().then(() => {
```

Comment:
So this can't use a `catch` because it will swallow any errors thrown in the `then`. I updated the code so that the tests pass and fail as expected but it is a little gross. It's mostly because the tests and the code under test are depending on the same promise instead of the tests depending on the promise that is returned by the code under test.